lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+        if (lc($ftr->{type}) eq "cds" || lc($ftr->{type}) eq "peg" ||
+                ($ftr->{type} eq "gene" and defined($ftr->{protein_translation}))) {
+            if (defined($ftr->{functions})){
+                $ftr->{function} = join("; ", @{$ftr->{functions}});
+            }
+            $oldfunchash->{$ftr->{id}} = $ftr->{function};
+            $ftr->{function} = "hypothetical protein";
+        }
+        elsif ($ftr->{type} eq "gene") {
+            $ftr->{type} = 'Non-coding '.$ftr->{type};
+        }
+        $oldtype->{$ftr->{id}} = $ftr->{type};
+        #
+        #   Count the input feature types
+        #
+        if (exists $types{$ftr->{type}}) {
+            $types{$ftr->{type}} += 1;
+        } else {
+            $types{$ftr->{type}} = 1;
+        }
+    }
+    if (exists $in_genome->{non_coding_features}) {
+        for (my $i=0; $i < @{$in_genome->{non_coding_features}}; $i++) {
+            my $ftr = $in_genome->{non_coding_features}->[$i];
+            if (!defined($ftr->{type})) {
+                $ftr->{type} = "Non-coding";
+            }
+            $oldtype->{$ftr->{id}} = $ftr->{type};
+            #
+            #   Count the input feature types
+            #
+            if (exists $types{"Non-coding ".$ftr->{type}}) {
+                $types{"Non-coding ".$ftr->{type}} += 1;
+            } else {
+                $types{"Non-coding ".$ftr->{type}} = 1;
+            }
+        }
+    } else {
+        $in_genome->{non_coding_features} = [];
+    }
+    return ($oldfunchash, $oldtype, \%types, $in_genome);
+}
+
+
+# When the input object_ref is referencing to a genome object
+sub _create_inputgenome_from_genome {
+    my ($self, $inputgenome, $input_obj_ref) = @_;
+
+    $inputgenome = $self->_get_genome($input_obj_ref);
+
+    my ($contigref, $contigobj, $contigID_hash);
+    if($inputgenome->{domain} !~ /Eukaryota|Plant/){
+        if (defined($inputgenome->{contigset_ref})) {
+            $contigref = $inputgenome->{contigset_ref};
+        } elsif (defined($inputgenome->{assembly_ref})) {
+            $contigref = $inputgenome->{assembly_ref};
+        }
+        ($contigobj, $contigID_hash) = $self->_get_contigs($contigref);
+    }
+    my ($oldfunchash, $oldtype, $types_ref);
+    my %types = ();
+    ($oldfunchash, $oldtype, $types_ref,
+        $inputgenome) = $self->_get_oldfunchash_oldtype_types($inputgenome);
+    %types = %{$types_ref};
+    return ($inputgenome, $contigobj, $contigID_hash, $oldfunchash, $oldtype, \%types);
+}
+
+
+# When the input_obj_ref is referencing to an assembly/contigset object
+sub _create_inputgenome_from_assembly {
+    my ($self, $inputgenome, $input_obj_ref) = @_;
+
+    $inputgenome->{assembly_ref} = $input_obj_ref;  # TOBe confirmed
+    my ($contigobj, $contigID_hash) = $self->_get_contigs($input_obj_ref);
+    return ($inputgenome, $contigobj, $contigID_hash);
+}
+
+sub _util_version {
+	my ($self) = @_;
+	my $VERSION = '0.1.6';
+    return $VERSION;
+}
```

Comment:
how does this relate to the version in the Impl file?

Reply:
This should be repeating what's in the Impl file because anno_utils.pm is the refactoring result of the main function in the Impl file.  The $VERSION variable should be a package variable in anno_utils.pm which is `use`d by the Impl. I made a slight change and will do more cleanup after the refactoring is officially accepted.